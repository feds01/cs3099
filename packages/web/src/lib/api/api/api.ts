/**
 * Generated by orval v6.6.4 üç∫
 * Do not edit manually.
 * Iamus API
 * This is a REST API for interfacing with Iamus. This API provides endpoints for interacting with user information, submissions, and reviews.
 * OpenAPI spec version: 1.0.0
 */
import type { GetVersion200, GetOpenapi200 } from '.././models';
import { customInstance } from '.././mutator/custom-instance';
import { useQuery, UseQueryOptions, QueryFunction, UseQueryResult, QueryKey } from 'react-query';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
type AsyncReturnType<T extends (...args: any) => Promise<any>> = T extends (...args: any) => Promise<infer R> ? R : any;

/**
 * Get the current version of the service
 * @summary Service version
 */
export const getVersion = () => {
    return customInstance<GetVersion200>({ url: `/version`, method: 'get' });
};

export const getGetVersionQueryKey = () => [`/version`];

export const useGetVersion = <TData = AsyncReturnType<typeof getVersion>, TError = unknown>(options?: {
    query?: UseQueryOptions<AsyncReturnType<typeof getVersion>, TError, TData>;
}): UseQueryResult<TData, TError> & { queryKey: QueryKey } => {
    const { query: queryOptions } = options || {};

    const queryKey = queryOptions?.queryKey ?? getGetVersionQueryKey();

    const queryFn: QueryFunction<AsyncReturnType<typeof getVersion>> = () => getVersion();

    const query = useQuery<AsyncReturnType<typeof getVersion>, TError, TData>(queryKey, queryFn, queryOptions);

    return {
        queryKey,
        ...query,
    };
};

/**
 * Get the current service OpenAPI schema
 * @summary Service OpenAPI schema
 */
export const getOpenapi = () => {
    return customInstance<GetOpenapi200>({ url: `/openapi`, method: 'get' });
};

export const getGetOpenapiQueryKey = () => [`/openapi`];

export const useGetOpenapi = <TData = AsyncReturnType<typeof getOpenapi>, TError = unknown>(options?: {
    query?: UseQueryOptions<AsyncReturnType<typeof getOpenapi>, TError, TData>;
}): UseQueryResult<TData, TError> & { queryKey: QueryKey } => {
    const { query: queryOptions } = options || {};

    const queryKey = queryOptions?.queryKey ?? getGetOpenapiQueryKey();

    const queryFn: QueryFunction<AsyncReturnType<typeof getOpenapi>> = () => getOpenapi();

    const query = useQuery<AsyncReturnType<typeof getOpenapi>, TError, TData>(queryKey, queryFn, queryOptions);

    return {
        queryKey,
        ...query,
    };
};
