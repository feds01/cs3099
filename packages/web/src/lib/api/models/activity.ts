/**
 * Generated by orval v6.6.4 üç∫
 * Do not edit manually.
 * Iamus API
 * This is a REST API for interfacing with Iamus. This API provides endpoints for interacting with user information, submissions, and reviews.
 * OpenAPI spec version: 1.0.0
 */
import type { ActivityKind } from './activityKind';
import type { ActivityReference } from './activityReference';
import type { ActivityType } from './activityType';
import type { User } from './user';

export interface Activity {
    id: string;
    owner: User;
    type: ActivityType;
    kind: ActivityKind;
    createdAt: number;
    updatedAt: number;
    message: string;
    references: ActivityReference[];
}
