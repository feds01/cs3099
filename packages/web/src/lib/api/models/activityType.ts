/**
 * Generated by orval v6.6.4 üç∫
 * Do not edit manually.
 * Iamus API
 * This is a REST API for interfacing with Iamus. This API provides endpoints for interacting with user information, submissions, and reviews.
 * OpenAPI spec version: 1.0.0
 */

export type ActivityType = 'user' | 'publication' | 'comment' | 'review';

// eslint-disable-next-line @typescript-eslint/no-redeclare
export const ActivityType = {
    user: 'user' as ActivityType,
    publication: 'publication' as ActivityType,
    comment: 'comment' as ActivityType,
    review: 'review' as ActivityType,
};
