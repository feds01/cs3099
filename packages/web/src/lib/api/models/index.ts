export * from './successStatus';
export * from './skipQuery';
export * from './flagSchema';
export * from './publicationActivityType';
export * from './successStatusObject';
export * from './takeQuery';
export * from './activityReference';
export * from './publicationActivity';
export * from './userActivity';
export * from './commentActivity';
export * from './reviewActivity';
export * from './commentActivityType';
export * from './multipleErrorMessage';
export * from './activityType';
export * from './responseError';
export * from './emailValidation';
export * from './userPatchRequest';
export * from './tokenRequest';
export * from './userRole';
export * from './userActivityType';
export * from './userLogin';
export * from './activityKind';
export * from './user';
export * from './userRegistration';
export * from './reviewActivityType';
export * from './followerCount';
export * from './publication';
export * from './uploadResource';
export * from './fileResponseType';
export * from './createPublicationRequest';
export * from './revisePublicationRequest';
export * from './directoryResponseType';
export * from './directoryResponseEntriesItem';
export * from './directoryResponseEntriesItemType';
export * from './fileResponse';
export * from './directoryResponse';
export * from './reviewStatus';
export * from './review';
export * from './comment';
export * from './patchComment';
export * from './sgUser';
export * from './commentAnchor';
export * from './sgPublication';
export * from './notification';
export * from './noContentResponse';
export * from './sgComment';
export * from './activity';
export * from './followingListData';
export * from './apiErrorResponseErrors';
export * from './apiErrorResponse';
export * from './patchPublicationRequest';
export * from './sgReview';
export * from './userAuthResponseResponse';
export * from './followerListResponse';
export * from './apiErrorResponseStatus';
export * from './createPublicationResponseResponse';
export * from './responseErrorMessage';
export * from './tokenVerificationResponseResponse';
export * from './resourceResponseResponse';
export * from './createCommentRequestBody';
export * from './getVersion200';
export * from './usernameValidation';
export * from './tokenResponseResponse';
export * from './resourceResponseResponseEntry';
export * from './getOpenapi200Schema';
export * from './getSgSsoLoginParams';
export * from './postSgResourcesImportParams';
export * from './postAuthEmailvalidity200';
export * from './postAuthSsoParams';
export * from './postAuthSso200';
export * from './postAuthUsernamevalidity200';
export * from './getSgSsoCallbackParams';
export * from './getUser200';
export * from './getUserParams';
export * from './getUserUsernameReviews200';
export * from './publicationMetadataResponseResponse';
export * from './postSgSsoVerifyParams';
export * from './patchUserUsername200';
export * from './getUserUsernameFollow200';
export * from './getUserUsername200';
export * from './getOpenapi200';
export * from './getUserUsernameFeedParams';
export * from './getUserUsernameFollowingParams';
export * from './getUserUsernameFollowersParams';
export * from './publicationExportResponseResponse';
export * from './getPublicationParams';
export * from './getPublicationUsername200';
export * from './followerListData';
export * from './getPublicationUsernameParams';
export * from './deletePublicationUsernameNameParams';
export * from './getUserUsernameFeed200';
export * from './patchUserUsernameRoleBody';
export * from './getPublication200';
export * from './getPublicationUsernameNameParams';
export * from './getPublicationUsernameNameRevisions200';
export * from './getPublicationUsernameNameTreePathParams';
export * from './getPublicationUsernameNameRevisionsParams';
export * from './postPublicationUsernameNameExportParams';
export * from './apiSuccessResponse';
export * from './getPublicationUsernameNameSources200';
export * from './deletePublicationUsernameNameAll200';
export * from './getPublicationUsernameNameReviewsParams';
export * from './postPublicationUsernameNameReview200';
export * from './postPublicationUsernameNameReviewParams';
export * from './getPublicationUsernameNameReviews200';
export * from './getPublicationbyidId200';
export * from './postPublicationbyidIdExportParams';
export * from './singleErrorMessage';
export * from './postPublicationbyidIdReview200';
export * from './postResourceUploadUsername200';
export * from './getPublicationbyidIdSources200';
export * from './getPublicationbyidIdReviewsParams';
export * from './getPublicationUsernameNameSourcesParams';
export * from './putReviewIdComment200';
export * from './getCommentId200';
export * from './getThreadId200';
export * from './getPublicationbyidIdReviews200';
export * from './patchPublicationUsernameNameParams';
export * from './getReview200';
export * from './patchCommentId200';
export * from './getSearchPublicationParams';
export * from './getSearchPublication200';
export * from './getActivityId200';
export * from './getReviewId200';
export * from './getSearchUserParams';
export * from './getActivityParams';
export * from './getNotificationsParams';
export * from './getActivity200';
export * from './getPublicationUsernameName200';
export * from './getReviewParams';
export * from './getNotifications200';
export * from './getSearchUser200';
export * from './getReviewIdComments200';
export * from './getUserUsernameRole200';
export * from './patchUserUsernameRole200';
export * from './getNotificationsId200';
export * from './postUserUsernameFollow200';
