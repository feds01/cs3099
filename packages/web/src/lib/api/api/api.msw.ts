/**
 * Generated by orval v6.6.4 🍺
 * Do not edit manually.
 * Iamus API
 * This is a REST API for interfacing with Iamus. This API provides endpoints for interacting with user information, submissions, and reviews.
 * OpenAPI spec version: 1.0.0
 */
import { SuccessStatus } from '.././models';
import faker from 'faker';
import { rest } from 'msw';

export const getGetVersionMock = () => ({
    status: faker.helpers.randomize(Object.values(SuccessStatus)),
    version: faker.random.word(),
});

export const getGetOpenapiMock = () => ({ status: faker.helpers.randomize(Object.values(SuccessStatus)), schema: {} });

export const getApiMSW = () => [
    rest.get('*/version', (_req, res, ctx) => {
        return res(ctx.delay(1000), ctx.status(200, 'Mocked status'), ctx.json(getGetVersionMock()));
    }),
    rest.get('*/openapi', (_req, res, ctx) => {
        return res(ctx.delay(1000), ctx.status(200, 'Mocked status'), ctx.json(getGetOpenapiMock()));
    }),
];
