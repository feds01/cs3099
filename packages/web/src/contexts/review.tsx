import { Comment, Review } from '../lib/api/models';
import { Permission } from '../lib/utils/roles';
import React from 'react';

/** Internal state of the ReviewProvider */
export interface ReviewState {
    /** The review document */
    review: Review;
    /** Any comments that are attached to the current review */
    comments: Comment[];
    /** The permissions that the current user has on the review */
    permission: Permission;
}

export interface ReviewDispatch {
    refetch: () => void;
}

export type ReviewContext = {
    state: ReviewState;
    dispatch: ReviewDispatch;
};

const ReviewStateContext = React.createContext<ReviewContext>({
    // @ts-ignore
    state: { comments: [], review: {}, permission: { modify: false, delete: false } },
    dispatch: {
        refetch: () => undefined,
    },
});

interface ReviewProviderProps {
    children: React.ReactNode;
    state: ReviewState;
    refetch: () => void;
}

export const ReviewProvider = ({ children, state, refetch }: ReviewProviderProps) => {
    return (
        <ReviewStateContext.Provider value={{ state, dispatch: { refetch } }}>{children}</ReviewStateContext.Provider>
    );
};

export function useReviewDispatch(): ReviewDispatch {
    const context = React.useContext(ReviewStateContext);

    if (context === undefined) {
        throw new Error('useReviewDispatch must be used within ReviewProvider');
    }

    return context.dispatch;
}

export function useReviewState(): ReviewState {
    const context = React.useContext(ReviewStateContext);

    if (context === undefined) {
        throw new Error('useReviewState must be used within ReviewProvider');
    }

    return context.state;
}
