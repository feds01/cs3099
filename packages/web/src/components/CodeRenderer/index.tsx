import { Review } from '../../lib/api/models';
import { CommentThread } from '../../lib/utils/comment';
import { coerceExtensionToLanguage, getExtension } from '../../lib/utils/file';
import CommentButton from '../CommentButton';
import CommentThreadRenderer from '../CommentThreadRenderer';
import Highlight from './PrismRenderer';
import { styled } from '@mui/material';
import Box from '@mui/material/Box';
import React, { ReactElement } from 'react';
import { useSelectionState } from '../../contexts/selection';

export const Wrapper = styled('div')`
    font-family: sans-serif;
    text-align: center;
`;

export const Pre = styled('pre')`
    text-align: left;
    margin: 1em 0;
    padding: 0.5em;
    overflow: scroll;

    & .token-line {
        line-height: 1.3em;
        height: 1.3em;
    }
`;

export const Line = styled('div')`
    display: table-row;
`;

export const LineNo = styled('span')`
    display: table-cell;
    text-align: right;
    padding-right: 1em;
    user-select: none;
    opacity: 0.5;
`;

export const LineContent = styled('span')`
    display: table-cell;
`;

interface CodeRendererProps {
    contents: string;
    filename: string;
    language?: string;
    review?: Review;
    commentMap?: Map<number, CommentThread[]>;
    lineNumbers?: boolean;
    lineOffset?: number;
    worker?: Worker;
}

export default function CodeRenderer({
    contents,
    filename,
    review,
    commentMap,
    language,
    lineNumbers = true,
    lineOffset = 0,
    worker,
}: CodeRendererProps): ReactElement {
    const extension = coerceExtensionToLanguage(getExtension(filename) ?? language ?? '');
    const selectionState = useSelectionState();

    return (
        <Highlight code={contents} language={extension ?? 'text'} worker={worker}>
            {({ style, tokens, getLineProps, getTokenProps }) => (
                <Pre style={style}>
                    {tokens.map((line, i) => (
                        <React.Fragment key={i}>
                            {typeof review !== 'undefined' && review.status === 'started' ? (
                                <CommentButton key={i} review={review} location={i} filename={filename}>
                                    <Line {...getLineProps({ line })}>
                                        {lineNumbers && <LineNo>{i + 1 + lineOffset}</LineNo>}
                                        <LineContent sx={{ userSelect: selectionState.isDragging ? 'none' : 'auto' }}>
                                            {line.map((token, key) => (
                                                <span key={key} {...getTokenProps({ token, key })} />
                                            ))}
                                        </LineContent>
                                    </Line>
                                </CommentButton>
                            ) : (
                                <Line {...getLineProps({ line })}>
                                    {lineNumbers && <LineNo>{i + 1 + lineOffset}</LineNo>}
                                    <LineContent>
                                        {line.map((token, key) => (
                                            <span key={key} {...getTokenProps({ token, key })} />
                                        ))}
                                    </LineContent>
                                </Line>
                            )}
                            {typeof review !== 'undefined' &&
                                commentMap?.get(i + 1)?.map((thread) => {
                                    return (
                                        <Box key={thread.id} sx={{ pt: 1, pb: 1 }}>
                                            <CommentThreadRenderer thread={thread} />
                                        </Box>
                                    );
                                })}
                        </React.Fragment>
                    ))}
                </Pre>
            )}
        </Highlight>
    );
}
