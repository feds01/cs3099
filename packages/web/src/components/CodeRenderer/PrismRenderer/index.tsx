import theme from './theme';
import { Component, ReactNode } from 'react';
import normalizeTokens from '../../../lib/utils/normalizeTokens';
import themeToDict, { type ThemeDict } from '../../../lib/utils/themeToDict';

import type {
    Language,
    Token,
    LineInputProps,
    LineOutputProps,
    TokenInputProps,
    TokenOutputProps,
    RenderProps,
    PrismGrammar,
    PrismToken,
} from '../../../types/renderer';

import { languages as prismLanguages, tokenize, hooks as prismHooks } from 'prismjs';

import 'prismjs/components/prism-clike';
import 'prismjs/components/prism-c';
import 'prismjs/components/prism-haskell';
import 'prismjs/components/prism-java';
import 'prismjs/components/prism-json';
import 'prismjs/components/prism-jsx';
import 'prismjs/components/prism-markdown';
import 'prismjs/components/prism-python';
import 'prismjs/components/prism-rust';
import 'prismjs/components/prism-tsx';
import 'prismjs/components/prism-typescript';
import 'prismjs/components/prism-v';
import 'prismjs/components/prism-yaml';
import 'prismjs/components/prism-markdown';
import 'prismjs/components/prism-toml';
import 'prismjs/components/prism-julia';
import 'prismjs/components/prism-core';

import { Box, CircularProgress } from '@mui/material';
import { HighlightWorkerResult } from '../../../worker/highlight.worker';
import { nanoid } from 'nanoid';
import { expr } from '../../../lib/utils/expr';

type Props = {
    language: Language;
    code: string;
    worker?: Worker;
    children: (props: RenderProps) => ReactNode;
};

interface HighlightState {
    loading: boolean;
    tokens: Token[][];
    nonce: string;
}

class Highlight extends Component<Props, HighlightState> {
    themeDict: ThemeDict;
    worker?: Worker;

    constructor(props: Props) {
        super(props);

        const grammar = prismLanguages[this.props.language];

        const tokens = expr(() => {
            if (typeof grammar === 'undefined') {
                return normalizeTokens([this.props.code] as unknown as (string | PrismToken)[]);
            } else {
                return [];
            }
        });

        this.state = {
            loading: tokens.length === 0,
            tokens,
            nonce: nanoid(),
        };

        // Assign the worker to the running instance
        this.worker = props.worker;

        // Create the theme...
        this.themeDict = themeToDict(theme, this.props.language);

        this.getLineProps = this.getLineProps.bind(this);
        this.getTokenProps = this.getTokenProps.bind(this);
        this.getStyleForToken = this.getStyleForToken.bind(this);
        this.handleWorkerError = this.handleWorkerError.bind(this);
        this.handleWorkerMessage = this.handleWorkerMessage.bind(this);
    }

    componentDidMount() {
        const grammar = prismLanguages[this.props.language];

        if (typeof this.worker !== 'undefined' && typeof grammar !== 'undefined') {
            this.worker.addEventListener('message', this.handleWorkerMessage);
            this.worker.addEventListener('error', this.handleWorkerError);

            this.worker.postMessage({ code: this.props.code, grammar, nonce: this.state.nonce });
        } else if (typeof grammar !== 'undefined') {
            const mixedTokens = grammar !== undefined ? this.tokenize(this.props.code, grammar) : [this.props.code];

            const tokens = normalizeTokens(mixedTokens as unknown as (string | PrismToken)[]);

            this.setState({
                loading: false,
                tokens,
            });
        }
    }

    componentWillUnmount() {
        if (typeof this.worker !== 'undefined') {
            this.worker.removeEventListener('message', this.handleWorkerMessage);
            this.worker.removeEventListener('error', this.handleWorkerError);
        }
    }

    handleWorkerMessage(event: MessageEvent<HighlightWorkerResult>) {
        if (event.data.nonce !== this.state.nonce) return;

        const { tokens } = event.data;

        this.setState({
            loading: false,
            tokens,
        });
    }

    handleWorkerError(event: ErrorEvent) {
        console.log(event);
    }

    getLineProps = ({ key, className, style, line, ...rest }: LineInputProps): LineOutputProps => {
        const output: LineOutputProps = {
            ...rest,
            className: 'token-line',
            style: undefined,
            key: undefined,
        };

        if (style !== undefined) {
            output.style = output.style !== undefined ? { ...output.style, ...style } : style;
        }

        if (key !== undefined) output.key = key;
        if (className) output.className += ` ${className}`;

        return output;
    };

    getStyleForToken = ({ types, empty }: Token) => {
        const typesSize = types.length;

        if (typesSize === 1 && types[0] === 'plain') {
            return empty ? { display: 'inline-block' } : undefined;
        } else if (typesSize === 1 && !empty) {
            return this.themeDict[types[0]];
        }

        const baseStyle = empty ? { display: 'inline-block' } : {};
        const typeStyles = types.map((type) => this.themeDict[type]);

        return Object.assign(baseStyle, ...typeStyles);
    };

    getTokenProps = ({ key, style, token, ...rest }: TokenInputProps): TokenOutputProps => {
        const output: TokenOutputProps = {
            ...rest,
            children: token.content,
            style: this.getStyleForToken(token),
            key: undefined,
        };

        if (typeof style !== 'undefined') {
            output.style = output.style !== undefined ? { ...output.style, ...style } : style;
        }

        if (key !== undefined) output.key = key;

        return output;
    };

    tokenize = (code: string, grammar: PrismGrammar): Array<Prism.Token | string> => {
        const env = {
            code,
            grammar,
            tokens: [] as (string | Prism.Token)[],
        };

        prismHooks.run('before-tokenize', env);
        const tokens = (env.tokens = tokenize(env.code, env.grammar));
        prismHooks.run('after-tokenize', env);

        return tokens;
    };

    render() {
        const { children } = this.props;
        const { tokens, loading } = this.state;

        if (loading) {
            return (
                <Box sx={{ display: 'flex', justifyContent: 'center' }}>
                    <CircularProgress />
                </Box>
            );
        }

        return children({
            tokens,
            style: this.themeDict.root,
            getLineProps: this.getLineProps,
            getTokenProps: this.getTokenProps,
        });
    }
}

export default Highlight;
