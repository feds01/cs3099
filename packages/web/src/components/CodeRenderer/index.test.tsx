import _ from 'lodash';
import CodeRenderer from '.';
import { ReviewStatus } from '../../lib/api/models';
import { mockReview } from '../../test-utils/mocks/review';
import renderWithWrapper from '../../test-utils/render';

jest.mock('../../contexts/auth');

describe('CodeRenderer tests', () => {
    const observeFn = jest.fn();
    const unobserveFn = jest.fn();

    afterEach(() => {
        jest.resetAllMocks();
    });

    it('CodeRender displays text with no highlighting', () => {
        const mockedReview = mockReview({ status: ReviewStatus.completed });

        // Mock the IntersectionObserver
        const mockIntersectionObserver = jest.fn();
        mockIntersectionObserver.mockReturnValue({
            observe: observeFn,
            unobserve: unobserveFn,
            disconnect: () => null,
        });
        window.IntersectionObserver = mockIntersectionObserver;

        const { asFragment } = renderWithWrapper(
            <CodeRenderer
                contents={'let x = 5;'}
                filename={'test.tsx'}
                lineNumbers={true}
                lineOffset={5}
                language={'typescript'}
                review={mockedReview}
            />,
        );

        // Test that the IntersectionObserver has been called once to observe the 'pre'
        // element that renders the code.
        expect(observeFn).toHaveBeenCalledTimes(1);

        expect(asFragment()).toMatchSnapshot();
    });

    it('CodeRender displays text with highlighting', () => {
        const mockedReview = mockReview({ status: ReviewStatus.completed });

        // This is a function to capture the callback that's passed into the intersection observer.
        // It is initialised with a dummy implementation which will then be later overridden with
        // the captured passed in callback once the IntersectionObserver is called.
        let intersectionCallback: (entries: Partial<IntersectionObserverEntry>[]) => void = jest.fn();

        // Mock the IntersectionObserver and capture the passed in callback from the intersection
        // observer.
        const mockIntersectionObserver = jest.fn().mockImplementation((callback, _options) => {
            intersectionCallback = callback;

            return {
                observe: observeFn,
                unobserve: unobserveFn,
                disconnect: () => null,
            };
        });

        window.IntersectionObserver = mockIntersectionObserver;

        const { asFragment, container } = renderWithWrapper(
            <CodeRenderer
                contents={'let x = 5;'}
                filename={'test.tsx'}
                lineNumbers={true}
                lineOffset={5}
                language={'typescript'}
                review={mockedReview}
            />,
        );

        const preElement = container.querySelector('pre')!;

        // Test that the IntersectionObserver has been called once to observe the 'pre'
        // element that renders the code.
        expect(observeFn).toHaveBeenCalledTimes(1);

        if (typeof intersectionCallback !== 'undefined') {
            intersectionCallback([
                {
                    isIntersecting: true,
                    target: preElement,
                },
            ]);
        }

        // Verify that we stop observing the element
        expect(unobserveFn).toHaveBeenCalledTimes(1);

        expect(asFragment()).toMatchSnapshot();
    });
});
