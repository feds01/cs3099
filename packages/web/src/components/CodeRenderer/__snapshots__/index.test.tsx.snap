// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CodeRenderer tests CodeRender displays text with highlighting 1`] = `
<DocumentFragment>
  <pre
    class="css-11fbl0n"
    style="color: rgb(57, 58, 52); background-color: rgb(246, 248, 250);"
  >
    <div
      class="token-line css-1j6x6lr"
    >
      <span
        class="css-8xmhzg"
      >
        6
      </span>
      <span
        class="css-f5iui0"
      >
        <span
          style="color: rgb(0, 0, 159);"
        >
          let
        </span>
        <span>
           x 
        </span>
        <span
          style="color: rgb(57, 58, 52);"
        >
          =
        </span>
        <span>
           
        </span>
        <span
          style="color: rgb(54, 172, 170);"
        >
          5
        </span>
        <span
          style="color: rgb(57, 58, 52);"
        >
          ;
        </span>
      </span>
    </div>
  </pre>
</DocumentFragment>
`;

exports[`CodeRenderer tests CodeRender displays text with no highlighting 1`] = `
<DocumentFragment>
  <pre
    class="css-11fbl0n"
    style="color: rgb(57, 58, 52); background-color: rgb(246, 248, 250);"
  >
    <div
      class="token-line css-1j6x6lr"
    >
      <span
        class="css-8xmhzg"
      >
        6
      </span>
      <span
        class="css-f5iui0"
      >
        <span>
          let x = 5;
        </span>
      </span>
    </div>
  </pre>
</DocumentFragment>
`;
